python/src/tasks/differential_expression.py

```diff
@@ -42,25 +69,25 @@ def compute(self):
         raw_adata = self.adata.raw.to_adata()
 
         # create a series to hold the conditions
-        raw_adata.obs["condition"] = None
+        raw_adata.obs["condition"] = ""
+
+        # mark all cells that aren't in the basis sample to be filtered out
+        self.mark_cells_not_in_basis_set(raw_adata, resp)
 
-        # fill in values appropriately. if `rest`, fill in all NaN
-        # values with the label, as the other one will be a cell set
-        # and override the appropriate values. if between two cell sets,
-        # make sure the cells not in either will be marked with `other`.
         for label, name in cell_sets.items():
-            if name == "rest" or "all" in name.lower():
-                raw_adata.obs["condition"].fillna(value=label, inplace=True)
+            if name == "background" or "all" in name.lower():
+                raw_adata.obs["condition"].loc[raw_adata.obs["condition"] == ""] = label
             else:
-                cells = find_cells_by_set_id(name, resp)
+                cells = self.get_cells_in_set(label, name, resp, cell_sets)
 
+                # Append the current label to the already existing one in "condition",
+                # this way when getting the cells for one cell set
+                # we can ignore those that are intersected with the other one
+                # since these would be problematic for the factor() function in the r worker.
                 raw_adata.obs["condition"].loc[
-                    raw_adata.obs["cell_ids"].isin(cells)
-                ] = label
-        
-        raw_adata.obs["condition"].fillna(value="other", inplace=True)
-
-        raw_adata.obs = raw_adata.obs.applymap(lambda x: x.decode() if isinstance(x, bytes) else x)
+                    (raw_adata.obs["cell_ids"].isin(cells))
+                    & (~raw_adata.obs["condition"].eq("FilteredOut"))
+                ] += label
```

Comment:
This does not handle the case that we discussed in Slack.

![image](https://user-images.githubusercontent.com/16019444/104755612-942f6180-5752-11eb-8637-3082e8dbecfd.png)

Reply:
I assume the helpers `get_all_cell_ids_on_current_level` and `find_cell_ids_in_same_hierarchy` are there to help you in implementing it.